internal/tracker/git.go

```diff
@@ -0,0 +1,170 @@
+// Copyright The Enterprise Contract Contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package tracker
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+	"sync"
+
+	gba "github.com/Maldris/go-billy-afero"
+	"github.com/go-git/go-git/v5"
+	"github.com/go-git/go-git/v5/plumbing/cache"
+	"github.com/go-git/go-git/v5/plumbing/object"
+	"github.com/go-git/go-git/v5/storage/filesystem"
+	"github.com/spf13/afero"
+
+	"github.com/enterprise-contract/ec-cli/internal/utils"
+)
+
+type gitTracker struct {
+	repositories *sync.Map
+}
+
+func NewGitTracker() *gitTracker {
+	g := gitTracker{}
+	g.repositories = &sync.Map{}
+
+	return &g
+}
+
+func (g gitTracker) Close(ctx context.Context) {
+	fs := utils.FS(ctx)
+
+	g.repositories.Range(func(_, val any) bool {
+		r, err := val.(func() (*git.Repository, error))()
+		if err != nil {
+			return true
+		}
+
+		bfs := r.Storer.(*filesystem.Storage).Filesystem()
+
+		// ignore error
+		_ = fs.RemoveAll(bfs.Root())
+
+		return true
+	})
+}
+
+func clone(ctx context.Context, repository string) (*git.Repository, error) {
```

Comment:
This assumes the default branch. We should document this somewhere. (At some point we may want to support other branches, but that seems like a premature feature at this point.)